Add new Torus Knot node to the set of generator nodes

DolphinDream opened the pull request:
Note: currently the node uses only single values from the input sockets.

This node is based on the blender’s torus knot + add-on, retrofitted to match sverchok node style.

![torusknotnode-example1](https://cloud.githubusercontent.com/assets/2083719/22677294/72f02470-ecbf-11e6-8bf2-69e5ba4d091c.png)
![torusknotnode-example2](https://cloud.githubusercontent.com/assets/2083719/22677297/72f6c0a0-ecbf-11e6-9e0a-3645392d0e8a.png)
![torusknotnode-example3](https://cloud.githubusercontent.com/assets/2083719/22677299/72f712bc-ecbf-11e6-83e9-3b8985e81f20.png)
![torusknotnode-example4](https://cloud.githubusercontent.com/assets/2083719/22677298/72f6dc7a-ecbf-11e6-90c2-3e326cba180a.png)
![torusknotnode-example5](https://cloud.githubusercontent.com/assets/2083719/22677296/72f6b02e-ecbf-11e6-81a3-44a934406c2b.png)
![torusknotnode-example6](https://cloud.githubusercontent.com/assets/2083719/22677295/72f6a764-ecbf-11e6-8ccd-a8a44ff55df1.png)

portnov commented:
Nice. 
But i think it should better be vectorized :)

zeffii commented on index.md:
```diff
@@ -22,6 +22,7 @@
     RandomVectorNodeMK2
     SvBricksNode
     ImageNode
+    TorusKnotNode
```
please postfix classnames using `Sv`

DolphinDream replied:
Sure. I can add that. What about the other nodes that don’t have Sv? (sphere, cylinder, plane etc) ?

zeffii replied:
Those are mostly old nodes, we have agreed new nodes will use that prefix.

zeffii commented:
if you need tips on `vectorization` let us know.

DolphinDream commented:
Yes I would like a hint on vectorization.

zeffii commented:
here's a few examples..
- https://github.com/nortikin/sverchok/blob/021979e956ffd5390fd658bd1caa1af8ca299a36/nodes/modifier_change/extrude_separate.py
- https://github.com/nortikin/sverchok/blob/86a37d523830b0f06a050b5967d8627915f2d963/nodes/generators_extended/box_rounded.py
- https://github.com/nortikin/sverchok/blob/021979e956ffd5390fd658bd1caa1af8ca299a36/nodes/generator/ngon.py

zeffii commented:
shall I close the other Pullrequest?

DolphinDream commented:
I added the Sv prefixes to both Torus and Torus Knot nodes classes. If you are referring to the other "Torus Node" pull request, sure you can go ahead and close it. I will work on an update to vectorize both nodes and make another pull request. Alternatively you could wait until I vectorize the nodes before you close the pull requests. It's up to you.

zeffii commented:
I didn't realize they were different nodes ( I should read better )

- consider combining them into one node

DolphinDream commented:
They are actually two different animals and I'm not sure combining them
into one node makes a whole lot of sense. One generates a curve (no mesh)
the other generates a mesh. Both are based on points that lie on a torus
surface though.

On Tue, Feb 7, 2017 at 9:18 AM, Dealga McArdle <notifications@github.com>
wrote:

> I didn't realize they were different nodes ( I should read better )
>
>    - consider combining them into one node
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nortikin/sverchok/pull/1209#issuecomment-278012473>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB_Lh9dw6_WidFB5pqbW9PdqfDyZHReuks5raH05gaJpZM4L5BjB>
> .
>



-- 
--
Marius Giurgi

zeffii commented:
we won't force you.

DolphinDream commented:
I'll give it a try and see how it goes  :)  i'll work on vectoring them
first.

On Tue, Feb 7, 2017 at 10:08 AM, Dealga McArdle <notifications@github.com>
wrote:

> we won't force you.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nortikin/sverchok/pull/1209#issuecomment-278027389>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB_Lh3jTQxMuT3G0BrJ0Dyd6HKYt-kjHks5raIjggaJpZM4L5BjB>
> .
>



-- 
--
Marius Giurgi

DolphinDream commented:
What's the preferred style to set the output sockets ?

    SvSetSocketAnyType(self, 'Vertices', [verts])

or

   self.outputs['Vertices'].sv_set(verts)



On Tue, Feb 7, 2017 at 10:08 AM, Dealga McArdle <notifications@github.com>
wrote:

> we won't force you.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nortikin/sverchok/pull/1209#issuecomment-278027389>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB_Lh3jTQxMuT3G0BrJ0Dyd6HKYt-kjHks5raIjggaJpZM4L5BjB>
> .
>



-- 
--
Marius Giurgi

zeffii commented:
` self.outputs['Vertices'].sv_set(verts)`  never the other.

zeffii commented:
I didn't realize that we still have many nodes that use `SvSetSocketAnyType`..

DolphinDream commented:
I'm almost done vectorizing the TorusNode but I need a little hint (my
python skills are rather rudimentary)

Here is the old non-vectorized code for vertices & normals:
if self.outputs['Vertices'].is_linked:
            verts, norms = torus_verts(R,r,n1,n2,rP,sP,self.Separate)
            SvSetSocketAnyType(self, 'Vertices', [verts])
            SvSetSocketAnyType(self, 'Normals', [norms])

where torus_verts returns both the verts list and norms list (it's faster
to compute the norms in the same path since it involves the same math which
can be computationally expensive)

Here is the current vectorized code for vertices ONLY:

if self.outputs['Vertices'].is_linked:
            verts = [torus_verts(aR,ar,an1,an2,arP,asP,self.Separate) for
aR, ar, an1, an2, arP, asP in zip(*parameters)]
            self.outputs['Vertices'].sv_set(verts)

where torus_verts returns just the verts list

The question is now how do i have the torus_verts still return the two
lists (verts  & norms) and have the vectorized code work fine with both
returned lists?

The next example obviously doesn't work, but this is how I envision it to
be like (with some additional zips or other list tricks):
if self.outputs['Vertices'].is_linked:
            verts, norms =
[torus_verts(aR,ar,an1,an2,arP,asP,self.Separate) for aR, ar, an1, an2,
arP, asP in zip(*parameters)]
            self.outputs['Vertices'].sv_set(verts)
            self.outputs['Normals'].sv_set(norms)


Note: It can be argued that the torus_verts should only return the verts
and not the norms and have a separate method just for the norms.. or even
let it to other Sverchock nodes to calculate the vertex norms. I just
thought it would be faster if the Torus node computes the normals itself
since the part of the math to calculate them is already done for the vertex
calculation.



On Tue, Feb 7, 2017 at 8:22 AM, Dealga McArdle <notifications@github.com>
wrote:

> here's a few examples..
>
>    - https://github.com/nortikin/sverchok/blob/
>    021979e956ffd5390fd658bd1caa1af8ca299a36/nodes/modifier_
>    change/extrude_separate.py
>    <https://github.com/nortikin/sverchok/blob/021979e956ffd5390fd658bd1caa1af8ca299a36/nodes/modifier_change/extrude_separate.py>
>    - https://github.com/nortikin/sverchok/blob/
>    86a37d523830b0f06a050b5967d8627915f2d963/nodes/generators_
>    extended/box_rounded.py
>    <https://github.com/nortikin/sverchok/blob/86a37d523830b0f06a050b5967d8627915f2d963/nodes/generators_extended/box_rounded.py>
>    - https://github.com/nortikin/sverchok/blob/
>    021979e956ffd5390fd658bd1caa1af8ca299a36/nodes/generator/ngon.py
>    <https://github.com/nortikin/sverchok/blob/021979e956ffd5390fd658bd1caa1af8ca299a36/nodes/generator/ngon.py>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nortikin/sverchok/pull/1209#issuecomment-277997967>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB_Lh8DiPzRMJb2gOA1vKS2d9pkLUe2qks5raHAVgaJpZM4L5BjB>
> .
>



-- 
--
Marius Giurgi

zeffii commented:
please use tripple backticks with the word python attached to the first set of backticks, to syntax highlight code

for example
```text

# ```python

# ```

```

zeffii commented:
if `torus_verts(arguments)` always returns `verts, norms` then a pattern we 've often used it to make a master list for each like `m_verts` and `m_norms` and append the output of `torus_verts` to each iteration

```python
m_verts = []
m_norams = []
for arguments in my_arg_list:
    v, n  = torus_verts(*arguments, self.Separate)
    m_verts.append(v)
    m_norms.append(n)

self.outputs['Vertices'].sv_set(m_verts)
self.outputs['Normals'].sv_set(m_norms)

```
understand this first before getting clever :)

zeffii commented:
note i moved the `self.Separate` to the front, so `arguments` can be 'unravelled' as real argments

DolphinDream commented:
Sure. this I already understand :) I just wanted to be clever ;) I'll do it
this way for readability. However, I'm curious how would you do it with
zips n'all.

On Tue, Feb 7, 2017 at 2:23 PM, Dealga McArdle <notifications@github.com>
wrote:

> if torus_verts(arguments) always returns verts, norms then a pattern we
> 've often used it to make a master list for each like m_verts and m_norms
> and append the output of torus_verts to each iteration
>
> m_verts = []
> m_norams = []for arguments in my_arg_list:
>     v, n  = torus_verts(*arguments, self.Separate)
>     m_verts.append(v)
>     m_norms.append(n)
> self.outputs['Vertices'].sv_set(m_verts)self.outputs['Normals'].sv_set(m_norms)
>
> understand this first before getting clever :)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nortikin/sverchok/pull/1209#issuecomment-278111781>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB_Lh-Z8GnTpYHPPD8yal-5JsfncsFodks5raMSZgaJpZM4L5BjB>
> .
>



-- 
--
Marius Giurgi

zeffii commented:
i think you mean this
```python
yip = [0, 2, 3, 4]
yap = [10, 20, 30, 40]

def my_yielder():
    for v, n in zip(yip, yap):
        yield(v, n)

m_generator = my_yielder()

verts, norms = zip(*m_generator)

print(verts)
print(norms)
```

zeffii commented:
probably something like this then
```python

def my_yielder():
    for arguments in my_arg_list:
        yield torus_verts(self.Separate, *arguments)

m_generator = my_yielder()
verts, norms = zip(*m_generator)

self.outputs['Vertices'].sv_set(m_verts)
self.outputs['Normals'].sv_set(m_norms)
```

DolphinDream commented:
Yea, I figured you need another zip.. 

This worked:

```python
        if self.outputs['Vertices'].is_linked:
            verts, norms = zip(*[torus_verts(aR,ar,an1,an2,arP,asP,self.Separate) for aR, ar, an1, an2, arP, asP in zip(*parameters)])
            self.outputs['Vertices'].sv_set(verts)
            self.outputs['Normals'].sv_set(norms)
```

however I implemented it like you suggested:

```python
        vertList=[]
        normList=[]
        if self.outputs['Vertices'].is_linked or self.outputs['Normals'].is_linked:
            for aR, ar, an1, an2, arP, asP in zip(*parameters):
                verts, norms = torus_verts(aR,ar,an1,an2,arP,asP,self.Separate)
                vertList.append(verts)
                normList.append(norms)
            self.outputs['Vertices'].sv_set(vertList)
            self.outputs['Normals'].sv_set(normList)
```

However, with both implementation when i use this vectorize node setup to draw the normals it seems that the normals are not connected correctly. Is there something wrong with the way i create the normal list or is my node setup not correct? (see screenshots for 1 torus vs 2 torii, driven by the range float values)

zeffii commented:
struggle with this yourself for a while, best way to learn.   ( also..dig through the other nodes)

zeffii commented:
the output of the UV connect is 1 list , not matter if your input is 1 nested list or 2 nested lists, I think you might have to `mesh join` the `vert+vert offset normal` before passing them to UV connect as a single stream.. not sure.

zeffii commented:
I think this has more to do with how UVConnect works than the output of your Node.

zeffii commented:
see similar: https://gist.github.com/7059ac99452793b7fe2ea3c3238b0112
![image](https://cloud.githubusercontent.com/assets/619340/22710144/87eb0d3c-ed7c-11e6-8849-6abd58d15986.png)

zeffii commented:
![image](https://cloud.githubusercontent.com/assets/619340/22710557/0c379dfc-ed7e-11e6-971f-112c4d31fe60.png)

zeffii commented on index.md:
```diff
@@ -22,6 +22,7 @@
     RandomVectorNodeMK2
     SvBricksNode
     ImageNode
+    SvTorusKnotNode
```
Also... this is probably destined for `Generators Extended` instead.

DolphinDream replied:
Yeah. you're right. I thought about but i wasn't sure. I'll make it ext in the next update once I vectorize it.

ly29 commented on nodes/generator/torusKnot.py:
```diff
@@ -0,0 +1,428 @@
+options_plus=True
+
+# greatest common denominator
+def gcd(a, b):
```
any reason not use ```math.gcd``` ? if so note in a comment.

DolphinDream replied:
Ha.. i didn't know python math has a gcd. Actually it seems that the 2.7
does not have it. I can remove it and use the math version.

DolphinDream commented:
Ha.. i didn't know python math has a gcd. Actually it seems that the 2.7
does not have it. I can remove it and use the math version.


On Wed, Feb 8, 2017 at 2:54 AM, ly29 <notifications@github.com> wrote:

> *@ly29* commented on this pull request.
> ------------------------------
>
> In nodes/generator/torusKnot.py
> <https://github.com/nortikin/sverchok/pull/1209#pullrequestreview-20699425>
> :
>
> > +from bpy.props import IntProperty, FloatProperty, BoolProperty, EnumProperty
> +
> +from math import sin, cos, pi, sqrt, radians
> +from mathutils import *
> +from random import random
> +import time
> +
> +from sverchok.node_tree import SverchCustomTreeNode
> +from sverchok.data_structure import updateNode, match_long_repeat, SvSetSocketAnyType
> +
> +DEBUG=False
> +
> +options_plus=True
> +
> +# greatest common denominator
> +def gcd(a, b):
>
> any reason not use math.gcd ? if so note in a comment.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nortikin/sverchok/pull/1209#pullrequestreview-20699425>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB_Lh_UMyioUD3uSHId8-K-uok6uKeYEks5raXTJgaJpZM4L5BjB>
> .
>



-- 
--
Marius Giurgi

DolphinDream commented:
TorusKnot node has been vector-eyezed ! :)

I also addressed the latest comments and cleaned up the code.

And here's the torus knot variety pack dozen, freshly baked :) 

![torusknotvectorized-example1](https://cloud.githubusercontent.com/assets/2083719/22742024/4fea9bf2-ede3-11e6-98b7-4cc38357cd6e.png)

DolphinDream commented:
As far as I'm concerned, if there are no other issues to be addressed, the request for Torus node can be closed.

ly29 commented:
Yeah if somebody who has time just tests is I approve.

zeffii commented:
@DolphinDream  nice work on the draw_buttons_ext btw.

zeffii commented on nodes/generator/torusKnot.py:
```diff
@@ -0,0 +1,432 @@
+            updateNode(self, context)
+
+    # GENERAL options
+    options_plus = BoolProperty(
```
indentation is 8 spaces deep here, i'd prefer to see 4 - but isn't a blocker.

DolphinDream replied:
I did not notice that :) Good catch. I will keep that in mind and perhaps change it to 4 in a future update.

zeffii commented on docs/nodes/generator/torusKnot.rst:
```diff
@@ -0,0 +1,91 @@
+Torus Knot
```
Looks good to me.

portnov commented:
For some reason if I just connect Torus Knot to "skin mesher", i have an error:

```
Error: Array length mismatch (expected 3, got 99)
TypeError: a float is required
  File "/home/portnov/.config/blender/2.78/scripts/addons/sverchok-master/core/update_system.py", line 322, in do_update_general
    node.process()
  File "/home/portnov/.config/blender/2.78/scripts/addons/sverchok-master/nodes/viz/viewer_skin_mk1.py", line 271, in process
    self.unit_generator(idx, geometry)
  File "/home/portnov/.config/blender/2.78/scripts/addons/sverchok-master/nodes/viz/viewer_skin_mk1.py", line 284, in unit_generator
    obj, data_layers = make_bmesh_geometry(self, bpy.context, geometry, idx, [radiix, radiiy])
  File "/home/portnov/.config/blender/2.78/scripts/addons/sverchok-master/nodes/viz/viewer_skin_mk1.py", line 119, in make_bmesh_geometry
    force_pydata(obj.data, verts, edges)
  File "/home/portnov/.config/blender/2.78/scripts/addons/sverchok-master/nodes/viz/viewer_skin_mk1.py", line 75, in force_pydata
    mesh.vertices.foreach_set('co', f_v)
Node Skin Mesher mk1b had exception couldn't access the py sequence
```

Not sure if it is problem in the mesher or in torus knot...

portnov commented:
Otherwise, it works
![output](https://cloud.githubusercontent.com/assets/284644/22747414/3a0958c8-ee51-11e6-98e2-10db94ca4cfb.gif)

DolphinDream commented:
Yeah. Apparently this is torus knot fault. I have corrected it and i will
make another push.

The issue may have been a double nested list [[]] instead of []. There is
currently a vectorization issue I was trying to solve with the knots that
are degenerate (Eg. p and q are not co-prime and thus they have disjoint
links that form one torus knot). I wanted these dijoint links to be
considered as ONE unit when replicating using vectorization. Instead,
because of the way the links are added to the list of vertices/edges etc
the links belonging to one torus knot are being considered separate and
thus replicated (via vectorization) in different locations (e..g when using
line or plane to replicate a torus knot). I knew about this issue and left
it to be resolved later. For now i'll revert the code so that the list are
not nested so that the skin mesher works fine.



On Wed, Feb 8, 2017 at 11:38 AM, Ilya V. Portnov <notifications@github.com>
wrote:

> For some reason if I just connect Torus Knot to "skin mesher", i have an
> error:
>
> Error: Array length mismatch (expected 3, got 99)
> TypeError: a float is required
>   File "/home/portnov/.config/blender/2.78/scripts/addons/sverchok-master/core/update_system.py", line 322, in do_update_general
>     node.process()
>   File "/home/portnov/.config/blender/2.78/scripts/addons/sverchok-master/nodes/viz/viewer_skin_mk1.py", line 271, in process
>     self.unit_generator(idx, geometry)
>   File "/home/portnov/.config/blender/2.78/scripts/addons/sverchok-master/nodes/viz/viewer_skin_mk1.py", line 284, in unit_generator
>     obj, data_layers = make_bmesh_geometry(self, bpy.context, geometry, idx, [radiix, radiiy])
>   File "/home/portnov/.config/blender/2.78/scripts/addons/sverchok-master/nodes/viz/viewer_skin_mk1.py", line 119, in make_bmesh_geometry
>     force_pydata(obj.data, verts, edges)
>   File "/home/portnov/.config/blender/2.78/scripts/addons/sverchok-master/nodes/viz/viewer_skin_mk1.py", line 75, in force_pydata
>     mesh.vertices.foreach_set('co', f_v)
> Node Skin Mesher mk1b had exception couldn't access the py sequence
>
> Not sure if it is problem in the mesher or in torus knot...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nortikin/sverchok/pull/1209#issuecomment-278382287>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB_Lh5cEFpnPO0GjXOaPSLdaRVGrQX9Wks5rae-YgaJpZM4L5BjB>
> .
>



-- 
--
Marius Giurgi

DolphinDream commented:
I pushed another update that should fix this. Let me know if it works for
you now.

On Wed, Feb 8, 2017 at 11:38 AM, Ilya V. Portnov <notifications@github.com>
wrote:

> For some reason if I just connect Torus Knot to "skin mesher", i have an
> error:
>
> Error: Array length mismatch (expected 3, got 99)
> TypeError: a float is required
>   File "/home/portnov/.config/blender/2.78/scripts/addons/sverchok-master/core/update_system.py", line 322, in do_update_general
>     node.process()
>   File "/home/portnov/.config/blender/2.78/scripts/addons/sverchok-master/nodes/viz/viewer_skin_mk1.py", line 271, in process
>     self.unit_generator(idx, geometry)
>   File "/home/portnov/.config/blender/2.78/scripts/addons/sverchok-master/nodes/viz/viewer_skin_mk1.py", line 284, in unit_generator
>     obj, data_layers = make_bmesh_geometry(self, bpy.context, geometry, idx, [radiix, radiiy])
>   File "/home/portnov/.config/blender/2.78/scripts/addons/sverchok-master/nodes/viz/viewer_skin_mk1.py", line 119, in make_bmesh_geometry
>     force_pydata(obj.data, verts, edges)
>   File "/home/portnov/.config/blender/2.78/scripts/addons/sverchok-master/nodes/viz/viewer_skin_mk1.py", line 75, in force_pydata
>     mesh.vertices.foreach_set('co', f_v)
> Node Skin Mesher mk1b had exception couldn't access the py sequence
>
> Not sure if it is problem in the mesher or in torus knot...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nortikin/sverchok/pull/1209#issuecomment-278382287>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB_Lh5cEFpnPO0GjXOaPSLdaRVGrQX9Wks5rae-YgaJpZM4L5BjB>
> .
>



-- 
--
Marius Giurgi

portnov commented:
Yes, now it works for me.

zeffii commented:
Congratulations @DolphinDream on your first node.

DolphinDream commented:
Thanks. I'm honored to have contributed to an already awesome add-on :) 

Btw, there is still one aspect about the Torus Knot node that needs to be resolved. I described it in the last commit message. It has to do with the degenerate torus knots (when p and q are not co-prime), which creates separate disconnected links. When using vectorization (e.g. a set of locations to replicate the torus knots, using a plane or a line for instance) the disconnected links of a degenerate torus replicate as individual meshes, instead of one disconnected mesh. I wonder if this issue is related to the way the TK node output is configured (some nested list format) or it has to do with the way the viewers interpret these lists. I would like some hint about how can I resolve this, whether within the node itself or via some node configuration.
